docs/rfcs/035-safekeeper-dynamic-membership-change.md

```diff
@@ -0,0 +1,457 @@
+from majority of `sk_set` and majority of `new_sk_set` (if it is present), it
+establishes this configuration as its own and moves to voting. 
+
+It should stop talking to safekeepers not listed in the configuration at this
+point, though it is not unsafe to continue doing so.
+
+To be elected it must receive votes from both majorites if `new_sk_set` is present.
+Similarly, to commit WAL it must receive flush acknowledge from both majorities.
+
+If walproposer hears from safekeeper configuration higher than his own (i.e.
+refusal to accept due to configuration change) it simply restarts.
+
+### Change algorithm
+
+The following algorithm can be executed anywhere having access to configuration
+storage and safekeepers. It is safe to interrupt / restart it and run multiple
+instances of it concurrently, though likely one of them won't make
+progress then. It accepts `desired_set: Vec<NodeId>` as input. 
+
+Algorithm will refuse to make the change if it encounters previous interrupted
+change attempt, but in this case it will try to finish it.
+
+It will eventually converge if old majority, new majority and configuration
+storage are reachable.
+
+1) Fetch current timeline configuration from the configuration storage.
+2) If it is already joint one and `new_set` is different from `desired_set`
+   refuse to change. However, assign join conf to (in memory) var
+   `join_conf` and proceed to step 4 to finish the ongoing change.
+3) Else, create joint `joint_conf: Configuration`: increment current conf number
+   `n` and put `desired_set` to `new_sk_set`. Persist it in the configuration
+   storage by doing CAS on the current generation: change happens only if
+   current configuration number is still `n`. Apart from guaranteeing uniqueness
+   of configurations, CAS linearizes them, ensuring that new configuration is
+   created only following the previous one when we know that the transition is
+   safe. Failed CAS aborts the procedure.
+4) Call `PUT` `configuration` on safekeepers from the current set,
+   delivering them `joint_conf`. Collecting responses from majority is required
+   to proceed. If any response returned generation higher than 
+   `joint_conf.generation`, abort (another switch raced us). Otherwise, choose
+   max `<last_log_term, flush_lsn>` among responses and establish it as 
+   (in memory) `sync_position`. We can't finish switch until majority 
+   of the new set catches up to this position because data before it 
+   could be committed without ack from the new set.
+4) Initialize timeline on safekeeper(s) from `new_sk_set` where it 
+   doesn't exist yet by doing `pull_timeline` from the majority of the 
+   current set. Doing  that on majority of `new_sk_set` is enough to
+   proceed, but it is reasonable to ensure that all `new_sk_set` members
+   are initialized -- if some of them are down why are we migrating there?
+5) Call `PUT` `configuration` on safekeepers from the new set,
```

Comment:
Does pull_timeline have a failure mode where it downloads only part of the a timeline's history?

If pull_timeline partially succeeded such that timelines exist but have incomplete data (i.e. fail the `sync_position` check), and there's no compute running, what will ensure that the timelines all catch up?  (it feels like we might need to retry pull_timeline until the LSN has fully caught up)

Reply:
pull_timeline copies the state it observers when it starts. The question whether we need compute to pass `sync_position` check is indeed very important. If, as described, we always first perform conf switch on the old majority and only then pull_timeline then on the happy path its result contains needed data, no need for catching up. In theory this might not always be the case; timeline can be created somehow earlier (then pull_timeline is skipped), e.g. procedure interrupted is in the middle after pull_timeline and we retry. Then yes, compute should recover new sk set majority enough for the check to pass which should be reasonably fast.

In theory you might imagine situation where such unsynchronized safekeepers exists but compute disappears before syncing them. Then we might get stuck, but this looks extremely unlikely and can be fixed by manually waking compute (or redoing pull_timeline manually). If we ever encounter this we can add code to request cplane to wake compute; point 3) in the additional considerations says about it.

Reply:
>In theory you might imagine situation where such unsynchronized safekeepers exists but compute disappears before syncing them. Then we might get stuck, but this looks extremely unlikely and can be fixed by manually waking compute

I'd prefer to eliminate this corner case so that the automated routine always works -- can we do that by putting a loop around pull_timeline and just calling it until it succeeds up to the `sync_position`?

That way we would have the nice property of the migration routine being totally independent of whether the compute is running

Reply:
I think probability of this is close enough to zero (and has manual fix) to initially ignore this issue. If we ever notice this or just later it is indeed possible to add guards around pull_timeline to do it safely even if timeline already exists, without risk of losing data: similarly to local timeline deletions it should go under term checks (i.e. we copy data to temp directory, but before atomic rename check that it won't put the log position backwards). This is more complicated, so I'd do it separately, if needed.

Reply:
I don't fully understand why it would be super rare, since the probability of a node being restarted during a pull_timeline is pretty high, but we can pick this conversation up later at the point of implementation.

Maybe it will become clearer when adding tests that kill safekeepers during migrations

Reply:
Just restart is ok, restarted compute will synchronize all safekeepers. Problem appears only when compute is shut down while safekeepers are not synchronized. But we shut down compute only after 5 minutes of inactivity (including absense of writes), which is a plenty of time to catch up. Moreover, we actually run sync-safekeepers on shutdown, which also would force synchronization.